server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -791,7 +795,21 @@ public boolean isRequireAlias() {
 
     @Override
     public Index getConcreteWriteIndex(IndexAbstraction ia, Metadata metadata) {
-        return ia.getWriteIndex(this, metadata);
+        if (DataStream.isFailureStoreEnabled() && writeToFailureStore) {
+            // TODO: Should this be a harder backstop than an assert statement?
+            assert ia.isDataStreamRelated()
+                : "Attempting to write a document to a failure store but the targeted index is not a data stream";
+            // Resolve write index and get parent data stream to handle the case of dealing with an alias
+            String defaultWriteIndexName = ia.getWriteIndex().getName();
+            DataStream dataStream = metadata.getIndicesLookup().get(defaultWriteIndexName).getParentDataStream();
+            // TODO: Should this be a harder backstop than an assert statement?
+            assert dataStream.getFailureIndices().size() > 0
+                : "Attempting to write a document to a failure store but the target data stream does not have one enabled";
```

Comment:
These assertions: Do we want them to be stronger checks that will trigger at runtime? I can imagine things going poorly if a failure document ends up somewhere it shouldn't